### PR TITLE
Allow references to interior mutable data behind a feature gate

### PR DESCRIPTION
**Pull request by oli-obk**
supercedes #80373 by simply not checking for interior mutability on borrows of locals that have `StorageDead` and thus can never be leaked to the final value of the constant

tracking issue: https://github.com/rust-lang/rust/issues/80384

r? @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
>  by simply not checking for interior mutability on borrows of locals that have StorageDead and thus can never be leaked to the final value of the constant

That's only a partial solution though, right? It means we cannot support any getter-style methods such as
```rust
fn fst_ref<T, U>(x: &(T, U)) -> &T { &x.0 }
```
(and same for `&mut` I guess)

So, I am not sure if this is worth it, since it seems rather restrictive. (I have no idea what your previous plan was in https://github.com/rust-lang/rust/pull/80373 since I hadn't gotten around to reading it before you closed it.)

**Comment by oli-obk** [comment]
> That's only a partial solution though, right? It means we cannot support any getter-style methods such as

Uh... no. If the feature gate is active, we only error on locals that have no `StorageDead`. The only locals that do not have `StorageDead` is uninhabited locals or locals in the trailing expression of a constant. This is the static analysis that goes in hand with the dynamic interior mutability behind reference check in validation. The only way a local can end up getting interned is by it being part of the trailing expression. So we only need to check `Rvalue::Ref` of locals that do not get erased before the end of the CTFE run, as all other pointers would become dangling, and we do not/cannot protect against dangling pointers anyway beyond the dynamic check.

**Comment by oli-obk** [comment]
Note that we can just outright not check anything in `const fn`, as the borrow checker will already do that for us. You're right though, as implemented there will be problems with (re)borrowing function arguments.

**Comment by RalfJung** [comment]
> Uh... no. If the feature gate is active, we only error on locals that have no StorageDead. The only locals that do not have StorageDead is uninhabited locals or locals in the trailing expression of a constant. This is the static analysis that goes in hand with the dynamic interior mutability behind reference check in validation. The only way a local can end up getting interned is by it being part of the trailing expression. So we only need to check Rvalue::Ref of locals that do not get erased before the end of the CTFE run, as all other pointers would become dangling, and we do not/cannot protect against dangling pointers anyway beyond the dynamic check.

I think I am lost, I have no idea what this means. Having no `StorageDead` can also arise because there was no `StorageLive`, either.

And how does the *referent* matter, isn't if the *reference* for which we care if it ends up in the final value? When a const has value `&var.field.field`, we care about whether the reference ends up in the final value, so how does looking at `var` tells us anything about that...?!? I am confused.^^

**Comment by RalfJung** [comment]
Oh, I think I get it -- if the *referent* has a `StorageDead`, then there simply is no way that it can still live during interning, so it cannot be part of the final value.

But I am a bit worried about drawing ourselves into a corner here, since it doesn't work for reborrowing getters.

**Comment by RalfJung** [comment]
That said, that is a concern for stabilization -- for experimentation, this approach seems fine.

I hope I'll have tome to review the code changes tomorrow.

**Comment by oli-obk** [comment]
> Having no `StorageDead` can also arise because there was no `StorageLive`, either.

Afaict, in constants, that's not an issue, and even if it were, all that would mean is that our check forbids more things than we want. That's ok, we can be more permissive later.



> And how does the _referent_ matter, isn't if the _reference_ for which we care if it ends up in the final value? When we see an expression `let _x = &var.field.field`, we care about whether the reference ends up in the final value, so how does looking at `var` tell us anything about that...?!? I am confused.^^
> 
> Unless you mean we are checking the `_x` (I haven't though that one though), but you keep saying "reference of locals that have no StorageDead", which to me means you are checking `var`, not `_x`.
> 
> Some examples could help. :)

Heh, ok, yea, I see the misunderstanding, and it is *exactly* the problem I had with building a static analyis for this until eddyb gave me the enlightening clue. Let's start with an example:

```rust
const FOO: &Cell<i32> = &Cell::new(42);
```

expands to something like the following MIR:

```rust
StorageLive(_2)
_2 = 42;
StorageLIve(_1)
_1 = Cell::new(42) -> bb2

bb2:
StorageDead(_2)
_0 = &_1
return
```

Now, you are right, `_0` has neither `StorageLive` nor `StorageDead`, as its storage is handled by the "caller" (at least in a normal function). Similarly function arguments have no storage markers. The problematic operation in the above example is the `&_1`, as that creates a `&'static Cell<i32>` within a constant. And we know it's `'static`, because of the escaping scopes rule: `_1`'s scope is `'static`, because it is a temporary in the trailing expression of a constant. Thus it also has no `StorageDead` marker. All this PR does is forbid `Rvalue::Ref` if the local (`_1` in our case) has no `StorageDead` markers and the local's type (`Cell::<i32>` in this case) has interior mutability. By the way MIR is built, we know that this can only happen for the escaping locals. Our dynamic checks for dangling pointers do the same thing, but dynamically. They check whether the allocation still exists at the end. The only way that can happen is if they do not have a `StorageDead` statement.

So with this new analysis, you either get a hard error because you tried to take a reference to a local that has interior mutability *and* lives forever, or you get a validation error later because you created a dangling pointer. We can talk about dangling pointers independently, they are entirely orthogonal to preventing references to interior mutable data in constants. You can create the easily without having to do anything related to (interior) mutability.

**Comment by oli-obk** [comment]
> But I am a bit worried about drawing ourselves into a corner here, since it doesn't work for reborrowing getters.

I added this to a test, and it works. We don't actually need to check for such borrows in `const fn` at all, because the borrow checker will already prevent you from returning references to local data (and there is no enlosing scope rule for `const fn`).

**Comment by RalfJung** [comment]
> I added this to a test, and it works. We don't actually need to check for such borrows in const fn at all, because the borrow checker will already prevent you from returning references to local data (and there is no enlosing scope rule for const fn).

Uh, I don't see the obvious argument for why this is sound.  The returned reference could come from various places, including things like `Box::leak`.

I also saw this in the code:
> A set that stores for each local whether it has a `StorageDead` for it somewhere.

My soundness argument above only works if `StorageDead` is *executed* on the referent in some block that post-dominates the block where the reference happens. It is nit enough for there to be a `StorageDead` just anywhere. So this seems suspicious to me.

**Comment by oli-obk** [comment]
> Uh, I don't see the obvious argument for why this is sound. The returned reference could come from various places, including things like `Box::leak`.

For a `const fn` by itself, this is a non-issue. It's totally fine for a `const fn` to produce `&'static Cell<i32>`. Of course, now you can create a `const` item and call such a `const fn` there. I am not sure yet how to handle this, and we will need to consider cases like this *anyway* when stabilizing const heap. But *right now* there is nothing that could possibly be problematic.

The only places other than a heap allocation that you can return references to is things that the function arguments point to (if function arguments are references or pointers), or static items. `const fn` cannot refer to static items, thus we are fine here. Anything that you can point to at the argument site must obviously be something at the caller site, and thus follows whatever rules apply to the caller. Either it's a `const` item, then it can't possibly be anything but a promoted value or a non-escaping local (as the enclosing scope rule does not apply to function arguments).

> The soundness argument above only works if `StorageDead` is _executed_ on the referent in some block that post-dominates the block where the reference happens. It is not enough for there to be a `StorageDead` just anywhere. So this seems suspicious to me.

I... don't know how we could have MIR like that. It's a valid concern that we can resolve by adding MIR validation checks I believe. It should never happen afaik.

We can additionally count the number of assignments to the local, and if there is only one assignment and only one `StorageDead`, the `StorageDead` must come after the assignment (by construction, not by possible values of `mir::Body`).

**Comment by RalfJung** [comment]
> For a const fn by itself, this is a non-issue. It's totally fine for a const fn to produce &'static Cell<i32>. Of course, now you can create a const item and call such a const fn there. I am not sure yet how to handle this, and we will need to consider cases like this anyway when stabilizing const heap. But right now there is nothing that could possibly be problematic.

I think there is something that could possibly be problematic: a `const` using such a `const fn` to put an `&Cell<i32>` in its result.

> The only places other than a heap allocation that you can return references to is things that the function arguments point to (if function arguments are references or pointers), or static items. const fn cannot refer to static items, thus we are fine here. Anything that you can point to at the argument site must obviously be something at the caller site, and thus follows whatever rules apply to the caller. Either it's a const item, then it can't possibly be anything but a promoted value or a non-escaping local (as the enclosing scope rule does not apply to function arguments).

Ah I see. That argument makes sense, but it very fundamentally relies on "no heap allocations". So if we stabilize this we say goodbye to some ways of supporting heap allocations -- IOW, we shouldn't stabilize this before having figured out the heap allocation story.

> We can additionally count the number of assignments to the local, and if there is only one assignment and only one StorageDead, the StorageDead must come after the assignment (by construction, not by possible values of mir::Body).

I don't follow this argument. "By construction" sounds very similar to "by the possible shapes of a MIR body" to me.

**Comment by oli-obk** [comment]
> I don't follow this argument. "By construction" sounds very similar to "by the possible shapes of a MIR body" to me.

Oop, I meant by mir building ("mir construction"), not the formal "by construction".



> we shouldn't stabilize this before having figured out the heap allocation story.

I can absolutely live with that. We should probably explore this feature with heap on nightly. Actually, this PR is necessary to even be able to experiment with generic heap abstractions, as that requires borrowing generic types.

**Comment by RalfJung** [comment]
> Oop, I meant by mir building ("mir construction"), not the formal "by construction".

Hm, that still seems like a rather fragile invariant.

At the very least, the code (and tracking issue) should explicitly spell out what we are relying on: if a `StorageDead` exists for a local, then for every assignment to the local, on each pass from that assignment to any `return`, there will be a `StorageDead` for this local.

**Comment by oli-obk** [comment]
> if a `StorageDead` exists for a local, then for every assignment to the local, on each pass from that assignment to any `return`, there will be a `StorageDead` for this local.

Added to the unresolved questions in the tracking issue.

**Comment by oli-obk** [comment]
> I think there is something that could possibly be problematic: a `const` using such a `const fn` to put an `&Cell<i32>` in its result

yes, but that's what I mean. *right now* there is no way to write such a `const fn`. So we need to find a solution that considers const heap and references to interior mutable data at the same time. Adding that to the tracking issues of both features right now.

File: compiler/rustc_error_codes/src/error_codes/E0492.md [hunk 1]
```diff
@@ -3,10 +3,11 @@ A borrow of a constant containing interior mutability was attempted.
 Erroneous code example:
 
 ```compile_fail,E0492
+#![feature(const_refs_to_cell)]
```
Comment (RalfJung) [inline compiler/rustc_error_codes/src/error_codes/E0492.md:6]: What is this about? Doesn't seem great to have a nightly-only example in the docs.

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 2]
```diff
@@ -208,9 +208,36 @@ impl NonConstOp for LiveDrop {
...
+        )
+    }
+}
+
 #[derive(Debug)]
 pub struct CellBorrow;
 impl NonConstOp for CellBorrow {
+    fn status_in_item(&self, ccx: &ConstCx<'_, '_>) -> Status {
+        match ccx.const_kind() {
+            // The borrow checker does a much better job at handling these than we do
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: Please add a reference to the discussion that justifies this -- in particular, some kind of FIXME noting that this is *not compatible* with having heap allocations in consts.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 3]
```diff
@@ -282,6 +287,27 @@ impl Validator<'mir, 'tcx> {
...
+    fn local_has_storage_dead(&mut self, local: Local) -> bool {
+        let ccx = self.ccx;
+        self.local_has_storage_dead
+            .get_or_insert_with(|| {
+                struct StorageDeads {
+                    locals: BitSet<Local>,
+                }
+                impl Visitor<'tcx> for StorageDeads {
+                    fn visit_statement(&mut self, stmt: &Statement<'tcx>, _: Location) {
+                        if let StatementKind::StorageDead(l) = stmt.kind {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:299]: So each time we check the table for a different local, we scan the entire MIR for all locals and then throw away the result except for this one local? How does this make any sense?

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: Also I find the control flow here extremely confusing. I think the `hir::ConstContext::ConstFn` check should be in validation.rs, not here.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 4]
```diff
@@ -556,7 +582,13 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 );
 
                 if borrowed_place_has_mut_interior {
-                    self.check_op(ops::CellBorrow);
+                    // Locals without StorageDead follow the "enclosing scope" rule, meaning
+                    // they are essentially anonymous static items themselves.
+                    if self.local_has_storage_dead(place.local) {
+                        self.check_op(ops::CellBorrowBehindRef);
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:588]: Why "BorrowBehindRef"? I cannot make sense of this name.

File: compiler/rustc_error_codes/src/error_codes/E0492.md (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_error_codes/src/error_codes/E0492.md:6]: It's the only way to trigger this error code now. Everything else causes a feature error

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 5]
```diff
@@ -556,7 +582,13 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 );
 
                 if borrowed_place_has_mut_interior {
-                    self.check_op(ops::CellBorrow);
+                    // Locals without StorageDead follow the "enclosing scope" rule, meaning
+                    // they are essentially anonymous static items themselves.
+                    if self.local_has_storage_dead(place.local) {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:587]: Somewhere around here there should be a note [like this](https://github.com/rust-lang/rust/pull/80418#issuecomment-752125143).

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: There are other `Qualif`s in this file where we do the check like this. The only reason for this function to take the `ccx` is to allow us to make this distinction.

File: src/test/ui/consts/std/cell.rs [hunk 6]
```diff
@@ -1,18 +1,26 @@
...
+const FOO_CONST: Wrap<*mut u32> = Wrap(Cell::new(42).as_ptr());
+//~^ ERROR encountered dangling pointer
 
 static FOO3: Wrap<Cell<u32>> = Wrap(Cell::new(42));
+const FOO3_CONST: Wrap<Cell<u32>> = Wrap(Cell::new(42));
+
 // ok
 static FOO4: Wrap<*mut u32> = Wrap(FOO3.0.as_ptr());
+const FOO4_CONST: Wrap<*mut u32> = Wrap(FOO3_CONST.0.as_ptr());
+//~^ ERROR encountered dangling pointer
```
Comment (RalfJung) [inline src/test/ui/consts/std/cell.rs:18]: The comment above says "ok" but there's an error; something is off here.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:299]: The `BitSet` is computed once and memoized. We use this pattern for a few other such checks. If there are no borrows, the `BitSet` is never created.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 4 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:588]: Uh... that's a naming scheme I never updated. I only fixed it in the `&mut` PR, not here, will do

File: src/test/ui/consts/std/cell.rs [hunk 7]
```diff
@@ -1,18 +1,26 @@
+#![feature(const_refs_to_cell)]
+
 use std::cell::*;
 
 // not ok, because this would create a silent constant with interior mutability.
 // the rules could be relaxed in the future
 static FOO: Wrap<*mut u32> = Wrap(Cell::new(42).as_ptr());
-//~^ ERROR cannot borrow a constant which may contain interior mutability
+//~^ ERROR encountered dangling pointer
```
Comment (RalfJung) [inline src/test/ui/consts/std/cell.rs:8]: Maybe add a comment explaining why this is the error that we get?

Presumably, `let x = Wrap(Cell::new(42).as_ptr());` would also dangle in regular code, right?

File: src/test/ui/consts/std/cell.rs (hunk 6 above)
Comment (oli-obk) [reply src/test/ui/consts/std/cell.rs:18]: The "ok" is for the `static`. I'll add a comment to each item explaining why it is ok or not

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: Yeah the control flow is already hard to follow without your PR, that's right.^^

File: src/test/ui/consts/std/cell.rs (hunk 7 above)
Comment (oli-obk) [reply src/test/ui/consts/std/cell.rs:8]: You don't even need `Wrap` in that case. [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bec8eea17f363fbea68d27cf210fbc7d) shows with miri/mir-dump that the pointer already dangles once `x` is initialized.

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: Checking the `const_kind` in `ops.rs` makes sense when the *stabilization status* is different in different contexts.

But it makes no sense when the fundamental soundness consideration is different, as in this case. All soundness considerations should be in the other file. Are there existing checks in this file that are about soundness, not just stabilization?

File: compiler/rustc_error_codes/src/error_codes/E0492.md (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0492.md:6]: So there's a feature error even for code that is still not accepted when the feature is enabled? Isn't that a bug?

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 3 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:299]: Oh, I got the brace nesting wrong... `Option` has a `get_or_insert_with` method?!?

File: compiler/rustc_error_codes/src/error_codes/E0492.md (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_error_codes/src/error_codes/E0492.md:6]: hmm... good point ^^ Well, tbh, we have this a lot of times. You get a feature error, but even with the feature the code won't compile. Just take any feature that is gated at the syntax level... you can still run into a typeck error later.

File: src/test/ui/consts/std/cell.rs (hunk 6 above)
Comment (RalfJung) [reply src/test/ui/consts/std/cell.rs:18]: Well, elsewhere in this file the comment is for the entire following "block" (separated by empty lines), so I used the same interpretation here. :)

File: src/test/ui/consts/std/cell.rs (hunk 7 above)
Comment (RalfJung) [reply src/test/ui/consts/std/cell.rs:8]: Yeah, this is the usual "`as_ptr` of a temporary" issue. It should just be explained that that's what happens.

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: The only other one that is about soundness is `StaticAccess`, which forbids accesses to statics except in other statics.

File: compiler/rustc_error_codes/src/error_codes/E0492.md (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0492.md:6]: I am surprised that this happens here though, given how fine-grained the const checking setup is.

Also, stable doesn't show the feature errors anyway, does it? So for a stable user this new example here is really rather confusing.

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: Yeah IMO that check should be in `validation.rs`.

File: compiler/rustc_error_codes/src/error_codes/E0492.md (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_error_codes/src/error_codes/E0492.md:6]: yea... I looked into it, and it just works if I remove the feature gate, this must be a leftover from the original dataflow version of this PR

File: compiler/rustc_mir/src/transform/check_consts/ops.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:232]: done

**Comment by oli-obk** [comment]
Rebased and addressed all comments

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 8]
```diff
@@ -209,6 +209,32 @@ impl NonConstOp for LiveDrop {
...
+            &ccx.tcx.sess.parse_sess,
+            sym::const_refs_to_cell,
+            span,
+            "cannot borrow here, since the borrowed element may contain interior mutability",
+        )
+    }
+}
+
+#[derive(Debug)]
+/// A borrow of a type that contains an `UnsafeCell` somewhere. The borrow escapes to
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:235]:
```suggestion
/// A borrow of a type that contains an `UnsafeCell` somewhere. The borrow might escape to
```

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 9]
```diff
@@ -556,7 +582,25 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 );
 
                 if borrowed_place_has_mut_interior {
-                    self.check_op(ops::CellBorrow);
+                    match self.const_kind() {
+                        // In a const fn all borrows are transient or point to the places given via
+                        // references in the arguments. The borrow checker guarantees that.
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:587]:
```suggestion
                        // references in the arguments (so we already checked them with
                        // TransientCellBorrow/CellBorrow as appropriate).
                        // The borrow checker guarantees that no new non-transient borrows are created.
```

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 10]
```diff
@@ -556,7 +582,25 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
...
+                        // NOTE: Once we have heap allocations during CTFE we need to figure out
+                        // how to prevent `const fn` to create long-lived allocations that point
+                        // to (interior) mutable memory.
+                        hir::ConstContext::ConstFn => self.check_op(ops::TransientCellBorrow),
+                        _ => {
+                            // Locals without StorageDead follow the "enclosing scope" rule, meaning
+                            // they are essentially anonymous static items themselves.
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:594]: The comment is focusing on the wrong case -- it should explain why locals *with* `StorageDead` are okay to treat as "definitely transient".

File: src/test/ui/feature-gate/feature-gate-const_refs_to_cell.rs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+// check-pass
+
+#![feature(const_refs_to_cell)]
+
+const FOO: () = {
+    let x = std::cell::Cell::new(42);
+    let y = &x;
+};
```
Comment (RalfJung) [inline src/test/ui/feature-gate/feature-gate-const_refs_to_cell.rs:8]: Which is the test ensuring that this is *not* accepted without the feature gate?

**Comment by RalfJung** [comment]
Sorry for the delay. Just got some comment nits. :)

File: src/test/ui/feature-gate/feature-gate-const_refs_to_cell.rs (hunk 11 above)
Comment (oli-obk) [reply src/test/ui/feature-gate/feature-gate-const_refs_to_cell.rs:8]: That would be `src/test/ui/consts/const-multi-ref.rs` (you can see it in this PR as its message slightly changed)

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 12]
```diff
@@ -209,6 +209,32 @@ impl NonConstOp for LiveDrop {
...
+        )
+    }
+}
+
+#[derive(Debug)]
+/// A borrow of a type that contains an `UnsafeCell` somewhere. The borrow might escape to
+/// the final value of the constant, and thus we cannot allow this (for now). We may allow
+/// it in the future for static items.
 pub struct CellBorrow;
 impl NonConstOp for CellBorrow {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:239]: Should the error for this ConstOp be adjusted to say something about "final value of the constant may not contain borrows with interior mutability" or so?

Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:239]: I update the message with `this borrow to an interior mutable value may end up in the final value of this constant` (`constant` is `const fn`, `static` or `constant` depending on where the message is emitted)

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 13]
```diff
@@ -242,8 +242,8 @@ impl NonConstOp for CellBorrow {
             ccx.tcx.sess,
             span,
             E0492,
-            "cannot borrow a constant which may contain \
-            interior mutability, create a static instead"
+            "this borrow to an interior mutable value may end up in the final value of this {}",
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:245]: Is it "borrow to", not "borrow of"?

File: src/test/ui/consts/partial_qualif.stderr [hunk 14]
```diff
@@ -1,4 +1,4 @@
-error[E0492]: cannot borrow a constant which may contain interior mutability, create a static instead
+error[E0492]: this borrow to an interior mutable value may end up in the final value of this constant
```
Comment (RalfJung) [inline src/test/ui/consts/partial_qualif.stderr:1]: I guess the obvious response one might have to this error message is "so what?". I do not have a proposal for a better wording though, so I am okay with going with this for now.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 15]
```diff
@@ -556,7 +582,30 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
...
+                        // NOTE: Once we have heap allocations during CTFE we need to figure out
+                        // how to prevent `const fn` to create long-lived allocations that point
+                        // to (interior) mutable memory.
+                        hir::ConstContext::ConstFn => self.check_op(ops::TransientCellBorrow),
+                        _ => {
+                            // Locals StorageDead are known to not leak to the final constant, and
+                            // it is thus inherently safe to permit such locals to have their
+                            // address taken as we can't end up with a reference to them in the
+                            // final value without creating a dangling pointer, which will cause
+                            // errors during validation.
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:599]:
```suggestion
                            // final value.
```
If the pointer dangles, it doesn't reference the interior mutable data either.^^

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 16]
```diff
@@ -556,7 +582,30 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
...
+                        // In a const fn all borrows are transient or point to the places given via
+                        // references in the arguments (so we already checked them with
+                        // TransientCellBorrow/CellBorrow as appropriate).
+                        // The borrow checker guarantees that no new non-transient borrows are created.
+                        // NOTE: Once we have heap allocations during CTFE we need to figure out
+                        // how to prevent `const fn` to create long-lived allocations that point
+                        // to (interior) mutable memory.
+                        hir::ConstContext::ConstFn => self.check_op(ops::TransientCellBorrow),
+                        _ => {
+                            // Locals StorageDead are known to not leak to the final constant, and
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:595]:
```suggestion
                            // Locals with StorageDead are definitely not referenced by the final constant, and
```

File: src/test/ui/consts/partial_qualif.stderr (hunk 14 above)
Comment (oli-obk) [reply src/test/ui/consts/partial_qualif.stderr:1]: yea I was considering adding another `note: ` to it, or at least use the `--explain` system. I'll do the latter now I guess

Comment (oli-obk) [reply src/test/ui/consts/partial_qualif.stderr]: How do you like the lastest version?

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 17]
```diff
@@ -238,13 +238,32 @@ impl NonConstOp for TransientCellBorrow {
...
+            err.help(
+                "To fix this, the value can be extracted to separate \
+                `static` and then referenced.",
+            );
+        }
+        if ccx.tcx.sess.teach(&err.get_code().unwrap()) {
+            err.note(
+                "A constant containing interior mutable data behind a reference can allow you
+                 to modify that data. This would make multiple uses of a constant to be able to
+                 see different values and allow one to escape the `Send` and `Sync` requirements
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:262]:
```suggestion
                 see different values and allow circumventing the `Send` and `Sync` requirements
```

File: compiler/rustc_mir/src/transform/check_consts/ops.rs [hunk 18]
```diff
@@ -238,13 +238,32 @@ impl NonConstOp for TransientCellBorrow {
...
-        )
+        );
+        err.span_label(
+            span,
+            format!("this borrow of an interior mutable value may end up in the final value"),
+        );
+        if let hir::ConstContext::Static(_) = ccx.const_kind() {
+            err.help(
+                "To fix this, the value can be extracted to separate \
+                `static` and then referenced.",
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/ops.rs:255]: Don't we usually use lower-case and no trailing `.`?

Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs:255]: oh... yea that was my mistake due to the `teach` messages actually starting with capitalization, and I moved this around a bit.

Comment (oli-obk) [reply compiler/rustc_mir/src/transform/check_consts/ops.rs]: fixed

**Comment by RalfJung** [comment]
Beautiful, thanks for taking care of my nitpicking. :)
@bors r+

**Comment by oli-obk** [comment]
@bors retry

I just ran the wasm tests locally and they passed, so I call spurious
